core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java

```diff
@@ -143,6 +147,45 @@ public void onShardFailedFailure(DiscoveryNode master, TransportException e) {
         assertTrue(failure.get());
     }
 
+    public void testTimeout() throws InterruptedException {
+        final String index = "test";
+
+        clusterService.setState(stateWithStartedPrimary(index, true, randomInt(5)));
+
+        String indexUUID = clusterService.state().metaData().index(index).getIndexUUID();
+
+        AtomicBoolean progress = new AtomicBoolean();
+        assert !progress.get();
```

Comment:
also, since we also set it in one place, after decrementing the latch and we only decrement the latch in one place, after checking progress - do we really need it?